### ISSUE TITLE
Inertia steals requests to API routes

### ISSUE THREAD

**Issue by flexchar** [laravel/jetstream#1238]
- Jetstream Version: 2.16.0
- Jetstream Stack: Inertia 0.6.9
- Uses Teams: yes
- Laravel Version: 9.49
- PHP Version: 8.2.1

### Description:

Assume the default `routes/api.php` file. It contains `/user` endpoint. If I set the endpoint HTTP method to any other than GET it will fail by being loosing its request to Inertia.

```php
// api.php
Route::middleware('auth:sanctum')->any('/user', function (Request $request) {
    return $request->user();
});
```

Now I use Postman to make request that I imported from my active session in the browser (so as to include the cookies, headers and all the required for Sanctum). 

So unless I make GET request and get the proper answer of my user object, I get redirected to either homepage `/` or to the location defined in the `Referer` header. This is where the magic begins and I've been banging my head against the table for days now. I cannot seem to understand why Inertia is touching this. 

To better illustrate, below is the log of Octane/Swoole server access log. 

I make exactly three requests since launching a fresh `octane:start` session. The same is with `artisan serve`. 

1. Note the request being GET and working fine. 
2. Note the request being POST (or any other verb than GET) and receiving 302 redirect to conversation controller (I dump `convo show` from controller's respective method). Here I have the header `Referer: http://tg.localhost/conversations/8474` present in the request.
3. Notice the redirect to `/` after removing `Referer` header.

![image](https://user-images.githubusercontent.com/9198896/216046233-5125270e-4790-4e99-9cb8-256638345ce2.png)

### Sanctum impact

Now if I navigate to `app\Http\Kernel.php` and comment out the Sanctum middleware from `api` group as such
```php
protected $middlewareGroups = [
        'web' => [
            \App\Http\Middleware\EncryptCookies::class,
            \Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse::class,
            \Illuminate\Session\Middleware\StartSession::class,
            \Laravel\Jetstream\Http\Middleware\AuthenticateSession::class,
            \Illuminate\View\Middleware\ShareErrorsFromSession::class,
            \App\Http\Middleware\VerifyCsrfToken::class,
            \Illuminate\Routing\Middleware\SubstituteBindings::class,
            \App\Http\Middleware\HandleInertiaRequests::class, // BTW! Tried commenting this out, also didn't change anything.
            \Illuminate\Http\Middleware\AddLinkHeadersForPreloadedAssets::class,
        ],

        'api' => [
            // \Laravel\Sanctum\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
            'throttle:api',
            \Illuminate\Routing\Middleware\SubstituteBindings::class,
        ],
    ];
```
I then can make a succesful request and get response in JSON saying: `{"message":"Unauthenticated."}`. It of course makes sense I'm unauthenticated since there is no token. I wish to use Sanctum first-party SPA logic. https://laravel.com/docs/9.x/sanctum#spa-authentication
![image](https://user-images.githubusercontent.com/9198896/216048021-20797251-6b48-4ee7-84a5-3cea78f143ad.png)

### Attempts

Tried clearing caches, restarting all containers, running curl from inside etc. I'm eager to debug and solve the problems myself before opening issues but I really seem to have stumbled upon ultimatum between Sanctum SPA (I need first-party API support) or Inertia and I ask for help. 

### Dump CURL request to display headers & cookies I've used in the example below. Importable directly to Postman/Insomnia.

<details>
  <summary>Expand</summary>

   ```bash
curl 'http://tg.localhost/api/message/194872/classify' -X POST -H 'User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/109.0' -H 'Accept: application/json' -H 'Accept-Language: en-US,da;q=0.7,lt;q=0.3' -H 'Accept-Encoding: gzip, deflate, br' -H 'Referer: http://tg.localhost/conversations/8474' -H 'Content-Type: application/json' -H 'X-Requested-With: XMLHttpRequest' -H 'X-CSRF-TOKEN: eyJpdiI6InFIT01sNGtQV3pvOHViZ1RubDN5eWc9PSIsInZhbHVlIjoiSmZ1RGkxL0tzVy8xZ0RwV3ljekZIZXZGcnJJTVRXZ1kzQmFTSEdpT3VCTngzd0dENjRIMmJCQWMvc2w0a2ttbEg3MGNyTU4vWHJEZjFYSFpMNVVWUUVITkhPaVY1bVFleVR1VXdsT0RoSGsxd0xIZG1RaVhxSkYwRVprM0o1bysiLCJtYWMiOiI2ZmViNWY2NmE3NTc1ZWM5YzgzZGVkNzY4MDBjY2U4OWY3NDYwOGJjMWQ5MmQwMTdjYjk0Mjg2YmJhZDkyZDlkIiwidGFnIjoiIn0%3D' -H 'Origin: http://tg.localhost' -H 'Connection: keep-alive' -H 'Cookie: remember_web_59ba36addc2b2f9401580f014c7f58ea4e30989d=eyJpdiI6ImM2VEJUMy9EWnRQckRtczJKWWNqV1E9PSIsInZhbHVlIjoicStnRjU4eHRlTk8rR1VpYU5JcFdnS0R3eEs0a0c3UHdsQkw0dERubkl2dkczUnZBeEVUaU80aDkyWThGMlVqUTc2b2lXVUhYZlNCZThsZzhzd290YUNuWnhHa3V4ZDQvd2M3WitvdENvaVRNb0htVHZ5QmpnSkNXQzB5aDZkTHYreHJBTmJteGw5ci96a1dYalVnWGN4elA4UFdSYnl2OE5tU2hQb0RXRVVpbzhlLzMzbTB1akx6TGZDc2hhZFpqd3docXlwVTd4ZHAvVUFtaFBFQ2o1UVNMdit5eXgvb3k0Qld2eUx1RE15ND0iLCJtYWMiOiIyYTZmNTE4NzJjNjgwZmI0MmI1ZmExZWIxYjEwMjMwZTU1ZTY0ZjQzZTQ4NzQ4MWU3OWJlYTUxN2M5ZjYxZTAwIiwidGFnIjoiIn0%3D; XSRF-TOKEN=eyJpdiI6InFIT01sNGtQV3pvOHViZ1RubDN5eWc9PSIsInZhbHVlIjoiSmZ1RGkxL0tzVy8xZ0RwV3ljekZIZXZGcnJJTVRXZ1kzQmFTSEdpT3VCTngzd0dENjRIMmJCQWMvc2w0a2ttbEg3MGNyTU4vWHJEZjFYSFpMNVVWUUVITkhPaVY1bVFleVR1VXdsT0RoSGsxd0xIZG1RaVhxSkYwRVprM0o1bysiLCJtYWMiOiI2ZmViNWY2NmE3NTc1ZWM5YzgzZGVkNzY4MDBjY2U4OWY3NDYwOGJjMWQ5MmQwMTdjYjk0Mjg2YmJhZDkyZDlkIiwidGFnIjoiIn0%3D; local_tg_app_session=eyJpdiI6InRjQlFQWnZPbTh0NzhNdUJHeFZNT2c9PSIsInZhbHVlIjoiTjJyeVY1Q1AxWWFZQ1VMZWpVVGRaTC9JalBJSTQ3WDQzRkVpS0t6akpMaFhsckxpWWYvc3BIanNRdGFpaE1ndjFDdmk1V0hMR3hveUFXb3Ewd1A0NEk0bUJFNWtXQnFBSHhwbSt2SDN3a1N3bG9yRnc1WVZGUTh4L3lRc281K2UiLCJtYWMiOiJlMGY0OGQ1NTIzNjc3MTIyNDI3N2YzNjZkZGFiZWZkNmE4MTExYzU1M2I0ZWUxZDI1NWY2Yzk5MWU3YmZkZmEwIiwidGFnIjoiIn0%3D' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin' -H 'DNT: 1' -H 'Sec-GPC: 1' -H 'Pragma: no-cache' -H 'Cache-Control: no-cache' --data-raw '{"tag":"Positive"}'

   ```
</details>

**Comment by taylorotwell** [comment]
I'm having a hard time figuring out what exactly your end goal is. Forget Postman for a second. What are you trying to do with Sanctum and Inertia, etc.? Describe your application and how you want it to work. How will it be used without Postman?

**Comment by jessarcher** [comment]
I'm a bit confused with the relevance of the Octane, Postman, and cookie stuff so I'm just trying to isolate everything down to what appears to be the core of the issue:

> Assume the default routes/api.php file. It contains /user endpoint. If I set the endpoint HTTP method to any other than GET it will fail by being loosing its request to Inertia.

In a fresh install with the latest Laravel and Jetstream Inertia (with teams), I have made the following change to `routes/api.php`:

```diff
- Route::middleware('auth:sanctum')->get('/user', function (Request $request) {
+ Route::middleware('auth:sanctum')->any('/user', function (Request $request) {
      return $request->user();
  });
```

I've then spun up a server just using `php artisan serve` for simplicity, as it doesn't appear from your description that Octane is a contributing factor.

If I then make both `GET` and `POST` requests to `/api/user` I get the same "Unauthenticated" response:

![image](https://user-images.githubusercontent.com/4977161/216491541-99b29a3f-c582-4d6e-8d25-17b32fcb03a6.png)

The `\Laravel\Sanctum\Http\Middleware\EnsureFrontendRequestsAreStateful` middleware in the `api` group is commented by default, but it seems relevant from your description so I have tried with it uncommented as well, but with the same result. 

Are there any other vital steps I'm missing to replicate this in a fresh install?

**Comment by flexchar** [comment]
I'd like to apologize for false positive. It turns out that I learned a great lesson creating a fresh example repository. 

There were two mistakes at the core of this issue that I'm to be blamed for. 

1. When I configured Inertia [error handling](https://inertiajs.com/error-handling), I've changed to always render Inertia page on error:
```php
 public function render($request, Throwable $e)
    {
        $response = parent::render($request, $e);

        if (app()->environment(['testing'])) {
            return $response;
        }
        
        if (in_array($response->getStatusCode(), [404, 403])) {
            return Inertia::render('Error', [
                'status' => $response->getStatusCode(),
            ])
                ->toResponse($request)
                ->setStatusCode($response->getStatusCode());
        } elseif ($response->getStatusCode() === 419) {
            return back()->with([
                'message' => 'The page expired, please try again.',
            ]);
        }

        return $response;
    }
```
As such, this would cause Inertia redirect to the "Error" component on every error. My mistake for not addressing `$request->wantsJson()` scenario. Specifically note the `return back()` part.

2. I'm driven a lot be performance, and thus, I replaced `axios` provided by laravel with native `fetch` API. My code fetching `XSRF-TOKEN` cookie made mistake and encoded the last character in the token that is `=` to `%3D` thus causing CSRF mismatch error. However I never got to see that error because of the point 1.

This really inspires and shows me the value of creating an isolated repository. 

And I really appreciate the support from you. It's a great honor to use the software from Laravel ecosystem and see the core team growing. Thank you.
